### DISCUSSION TITLE
Catching exceptions when composing a json

### DISCUSSION THREAD

**Discussion by tawmoto** [nlohmann/json#3441] [Q&A] [answered]
Hi,

I read this page: https://json.nlohmann.me/home/exceptions/
And I am not sure if I should catch an exception when I compose a Json or not. (I am catching when I am parsing one)
Would this be useful in any way? Or it's just STL data that can throw?

```
try
{
  json j2 = {
    {"pi", 3.141},
    {"happy", true},
  };
}
catch(const Json::exception& ex)
{...}
```

Thank you

**Comment by falbrechtskirchinger** [comment, marked as answer]
If you look at the API documentation, for each function you'll find a list of the exceptions they may throw. For example: https://json.nlohmann.me/api/basic_json/basic_json/#exceptions

Of course that is on top of any exceptions the STL containers may throw. (Catching `json::exception` won't help there, but you knew that.)

To answer your specific question: the code you wrote should never catch an exception according to the documentation.

**Reply by tawmoto** [reply]
@falbrechtskirchinger thank you for your answer, so from what I see here: https://json.nlohmann.me/api/basic_json/basic_json/
I am using the 3rd ctor, which is `noexcept`, isn't it?

Thanks

**Reply by falbrechtskirchinger** [reply]
> @falbrechtskirchinger thank you for your answer, so from what I see here: https://json.nlohmann.me/api/basic_json/basic_json/ I am using the 3rd ctor, which is `noexcept`, isn't it?

No. You're using number 5. The one which takes an `initializer_list` and can only throw when the second parameter is set to `false`.

**Reply by tawmoto** [reply]
@falbrechtskirchinger great.
If I may, I have one last question, and when using this form?
```
Json json;
json["new_key"] = "some_string";
```
I use that form also in my code.

Thanks again. :-)

**Reply by nlohmann** [reply]
See https://json.nlohmann.me/api/basic_json/operator%5B%5D/#exceptions.

**Reply by falbrechtskirchinger** [reply]
> @falbrechtskirchinger great. If I may, I have one last question, and when using this form?
> 
> ```
> Json json;
> json["new_key"] = "some_string";
> ```
> 
> I use that form also in my code.

The relevant pages of the documentation are:
https://json.nlohmann.me/api/basic_json/operator%5B%5D/#exceptions
and
https://json.nlohmann.me/api/basic_json/operator=/

On the last one I'm not entirely sure if the `noexcept` condition evaluates to `true`.

But `operator[]` can throw for multiple different reasons as you can see. For example, if `json` is not an object the access attempt will throw.
